Revert "[New] support enumerable Symbol properties"

MartyJRE opened the pull request:
This reverts commit 7d659e78d43e69e6afc604e12b3ba3a5def0e46f.

The reason for this is that the new feature is a breaking change and should be versioned as such.

Fixes #5.

ljharb reviewed (changes requested):
A naive revert isn't the best solution here; it'd be to put the symbol part behind an option (ie, `(includeSymbols ? ownEnumerableKeys : objectKeys)` everywhere `ownEnumerableKeys` is currently called)

MartyJRE commented:
@ljharb I added options to traverse. It now only deals with string keys by default. When given `includeSymbols`: true inside the options, it functions the way you desgined.

Also added `immutable` to options to avoid having too many parameters.

ljharb commented on index.js:
```diff
@@ -49,12 +49,25 @@ function ownEnumerableKeys(obj) {
+ * Given includeSymbols, `ownEnumerableKeys` is used and iterates over symbols.
+ * Otherwise, `objectKeys` is used and iterates over strings only.
+ * @param {Object} options
+ * @returns {function (object: object): (string | symbol)[]}
```
```suggestion
 * @returns {typeof ownEnumerableKeys | typeof objectKeys}
```

ljharb commented on index.js:
```diff
@@ -141,9 +155,10 @@ function walk(root, cb, immutable) {
 		if (!alive) { return state; }
 
 		function updateState() {
+			var iteratorFunction = determineIterator(options);
```
this can be moved up to line 109.5

ljharb commented on index.js:
```diff
@@ -208,14 +223,27 @@ function walk(root, cb, immutable) {
 	var node = this.value;
-	for (var i = 0; i < ps.length; i++) {
-		var key = ps[i];
+	var psCopy = ps.slice();
+	if (!this.options.includeSymbols) {
+		psCopy = psCopy.filter(function (key) {
+			return typeof key !== 'symbol';
+		});
+	}
```
we can't depend on `filter` existing, and ideally not `slice` either.

ljharb commented on index.js:
```diff
@@ -248,11 +288,11 @@ Traverse.prototype.set = function (ps, value) {
 
 Traverse.prototype.map = function (cb) {
-	return walk(this.value, cb, true);
+	return walk(this.value, cb, Object.assign(this.options, { immutable: true }));
```
we definitely can't assume the existence of Object.assign.

ljharb commented:
(you'll also need to rebase; i pulled out the immutable → options refactor)

ljharb commented on .eslintrc:
```diff
@@ -19,7 +19,8 @@
 		"object-curly-newline": 0,
 		"sort-keys": 0,
-	},
+        "max-lines": "warn"
```
```suggestion
        "max-lines": "warn",
```

ljharb commented on index.js:
```diff
@@ -52,12 +52,54 @@ function ownEnumerableKeys(obj) {
+ * @param {Object} options
+ * @returns {typeof arr}
+ */
+function stripSymbols(arr, options) {
```
I’m confused, why would we need to strip them when we could avoid adding them in the first place?

ljharb commented:
Don’t worry about the automatic rebase workflow.

ljharb commented:
I've rebased this and condensed it down to one commit.

It's still failing 3 tests for me, and I'm not sure why.

ljharb commented:
@MartyJRE are you able to complete this PR? i'd like to get the fix released.
